### PR TITLE
feat: add metrics collection using Prometheus

### PR DESCRIPTION
**Pull request by jedel1043**
## Description

This change allows integrating the charm with the [Canonical Observability Stack](https://charmhub.io/cos-lite). Using Rivos Inc's [Prometheus exporter](https://github.com/rivosinc/prometheus-slurm-exporter), we can collect metrics from the controller, which can then be sent to a subordinate [grafana-agent](https://charmhub.io/grafana-agent) that can forward the metrics to the COS on Kubernetes using a cross-model, cross-cloud relation.

Still WIP:

- [ ] Tutorial on how to deploy the observability stack.
- [ ] Tests to check that everything deploys correctly.

## How was the code tested?

Tested locally (Ubuntu Noble) in a hybrid-cloud environment on Juju (Microk8s + LXD).

## Checklist

- [x] I am the author of these changes, or I have the rights to submit them.
- [x] I have added the relevant changes to the README and/or documentation.
- [x] I have self reviewed my own code.
- [x] All requested changes and/or review comments have been resolved.

### CODE REVIEW

**Comment by NucciTheBoss** [comment]
_Need to rebase experimental with main_

**Review by NucciTheBoss — CHANGES_REQUESTED** [review]
@jedel1043 can you rebase this branch with the latest changes from the `experimental` branch? You might need to resolve some conflicts.

I'm going to start working on putting together some test bundles that we can use within our documentation!

**Comment by jedel1043** [comment]
> @jedel1043 can you rebase this branch with the latest changes from the `experimental` branch? You might need to resolve some conflicts.

I'll do that after the CI is in a good state (probably after #31 gets merged).

**Comment by jedel1043** [comment]
Moment of truth...

**Comment by NucciTheBoss** [comment]
> ## How was the code tested?
> WIP

@jedel1043 could you update this section of the PR message to indicate how you verified that COS was working in a test deployment?

File: src/cos/grafana_dashboards/Slurm-overview.json
Comment (NucciTheBoss) [inline src/cos/grafana_dashboards/Slurm-overview.json:1]: [nit]: Is it possible to instead have this file name `slurm-overview.json`? Or is it required that this file name be capitalized?

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 ops==2.14.0
 distro==1.9.0
-pycryptodome==3.20.0
+pycryptodome==3.20.0
```
Comment (NucciTheBoss) [inline requirements.txt:3]: [nit]: Missing newline with the addition of `pycryptodome` as dependency. Weird that it's both adding and removing the line in diff?

[foresight]: we'll want to spend some time investigating next pulse using `poetry` for our charms after we reassemble the monorepo if we have the time.

```suggestion
pycryptodome==3.20.0

```

File: src/charm.py [hunk 2]
```diff
@@ -112,6 +120,8 @@ def _on_install(self, event: InstallEvent) -> None:
             self._legacy_manager.write_jwt_rsa(jwt_rsa)
 
             self._slurmctld.enable()
+            self._slurmctld.munge.enable()
+            self._slurmctld.exporter.enable()
```
Comment (NucciTheBoss) [inline src/charm.py:124]: [foresight]: we'll want to make the port the prometheus exporter uses configurable in the future as we'll potentially have multiple exporters running per-node. We should likely address this after we have a better idea of what exporters we'll need to have installed on each of the nodes.

**Review by NucciTheBoss — CHANGES_REQUESTED** [review]
Honestly, I'm surprised how painless the COS integration itself was into the charm. Most of the fun came from just getting the exporter supported :sweat_smile: 

Just two small nits I'd like you to address, and after that I'm good to ship :ship:

File: src/cos/grafana_dashboards/Slurm-overview.json
Comment (jedel1043) [reply src/cos/grafana_dashboards/Slurm-overview.json:1]: Fixed

File: requirements.txt (hunk 1 above)
Comment (jedel1043) [reply requirements.txt]: Added the newline

**Review by NucciTheBoss — APPROVED** [review]
LGTM :rocket:

**Pull request merged**